cli/src/ui.rs

```
@@ -81,10 +81,18 @@ impl UiOutput {
         Ok(UiOutput::Paged { child, child_stdin })
     }
 
-    fn new_builtin_paged() -> streampager::Result<UiOutput> {
+    fn new_builtin_paged(config: &StreampagerConfig) -> streampager::Result<UiOutput> {
+        // This uselessly reads ~/.config/streampager/streampager.toml, even
+        // though we then override the important options. TODO(ilyagr): Fix this
+        // once/if https://github.com/facebook/sapling/pull/1011 is merged.
         let mut pager = streampager::Pager::new_using_stdio()?;
```

I think it still make sense to load the config file. There may be key bindings, etc.?

---

I would prefer to copy any such config into `jj`, especially since streampager's current docs at https://github.com/markbt/streampager are both incomplete and seem to be incorrect (as we discovered in the other thread, setting `streampager.toml` to the values shown there didn't seem to work).

It seems difficult at the moment to improve streampager's own configuration and docs. If the situation changes later, and if more programs start using streampager, we could change our mind.

---

> I would prefer to copy any such config into jj, 

In that case, I would add config table specific to "streampager", not generic "builtin". It's unlikely all config knobs can be generalized.

I don't have a strong opinion about that. The current config loading at streampager side is broken, but we can send a fix to upstream. OTOH, if we fixed the upstream config deserialization implementation, maybe we can reuse it to load parameters from our TOML table.

---

I changed the key to `ui.streampager`. Perhaps we should also rename the pager to `:builtin-streampager`, with `:builtin` being an alias.

>  OTOH, if we fixed the upstream config deserialization implementation, maybe we can reuse it to load parameters from our TOML table.

That's what I was thinking.

---

> I changed the key to `ui.streampager`. Perhaps we should also rename the pager to `:builtin-streampager`, with `:builtin` being an alias.

I have no strong opinion about that. It's unlikely we would add more than one builtin pagers. FWIW, the syntax is `:<internal-name>`, so the canonical name would be `:streampager`.

> > OTOH, if we fixed the upstream config deserialization implementation, maybe we can reuse it to load parameters from our TOML table.
> 
> That's what I was thinking.

In that case, shouldn't we follow the upstream naming convention (e.g. `streampager.interface_mode = "delayed"`)? Otherwise, we'll need to reimplement deserializer.

https://github.com/markbt/streampager?tab=readme-ov-file#configuration

---

I was thinking of doing it only in those cases where Streampager's names seem good for us. For example, keybindings maybe.

Re Streampager's names: the names "interface_mode", "delayed", and "hybrid" all seem unclear to me. What's delayed? Is "hybrid" more hybrid than "delayed"? (Also, to explain "hybrid", you have to explain the "direct" mode, which is otherwise pretty useless) And `interface_mode` could be a setting for changing anything at all.

Let me know if you prefer Streampager's choices, or can think of something better. I'm not 100% happy with my choices, but together with the text in the documentation, I'm hoping they will be easy for people to wrap their head around.

---

>  doing it only in those cases where Streampager's names seem good for us. For example, keybindings maybe.

Ok.

> Re Streampager's names:

I agree it's unclear, but I think it's more correct than `clear-screen = never|always|if-long-or-slow`. It's hybrid in a way that the screen is not cleared *and* the pager exits immediately depending on the content length. (BTW, this part is what I don't like about streampager. I just want to turn off the screen reset as `less -RX` would do.)

---

I thought `clear-screen="never"` is better than `full-screen="never"`, because, for long files, you do get a full-screen interface even if you choose "never" (but the screen is not cleared).

The way I think of "clear-screen" is that it determines when the screen is cleared, and then there is the additional behavior that whenever the screen is *not* cleared, streampager exits on short inputs. I tried explaining this in config.md, I'm not sure whether it's enough to get people to not get confused.

I find `echo aa | LESS= less -RX` a bit weird, especially if you press `up` and `q` after it, which leaves a ton of `~`s on the screen. So, I sort of understand why streampager chose its set of tradeoffs instead of less's, and from this perspective behaving as I described in the previous paragraph sort-of makes sense to me. From other perspectives, it's a headache.

---

I misunderstood "never" would be mapped to `Direct`. So, yes, it's kinda `full-screen="if-long"`, and there's no "never".

To me, "clear/reset screen" sounds tightly related to terminal reset behavior, and I wouldn't expect it affects the interactive UI. Maybe it can be `interactive=always|if-long|if-long-or-slow` or the inverse `auto-quit=never|if-one-screen|if-one-screen-and-quick`?

---

I like these better, but they do not indicate the fact that "always" always clears the screen on exit, "if-long" never clears it on exit, and "if-long-or-slow" clears it only if the pager becomes interactive. When expressed in this way, I find it hard to wrap my head around.

There might be some way to present this information in a memorable way, similar to my attempt described in https://github.com/jj-vcs/jj/pull/5415#discussion_r1934515205 (that relied on focusing the user's attention on whether the screen is cleared). Then, I'd be happy to switch. I don't have it in mind right now, but we have a few days to think about it.

One way that comes to mind, but is probably not great, is `interactive=always|if-long-and-never-clear-screen|if-long-or-slow`.

---

> but they do not indicate the fact that "always" always clears the screen on exit,

Correct. I just assumed UI difference is more important to users than screen reset.

> "if-long" never clears it on exit, and "if-long-or-slow" clears it only if the pager becomes interactive.

Oh, weird. Maybe better to provide two separate flags then (and error out on unsupported combination)?

(I also think copying the upstream `interface-mode` is okay given that this flag is super abstract and not possible to describe in one word.)

---

I've also been thinking about providing two flags, but I'm leaning against it. It would make the delay flag hard to name. Here's what I'm getting for this option (that I don't really like):

```toml
[ui.streampager]
quit-if-one-screen = true  # default
clear-screen-on-exit = false # default false, has no effect if quit-if-one-screen = false

# Only relevant if quit-if-one-screen and clear-screen-on-exit are both true
quit-if-one-screen-delay-millis = 2000 # (defaults to 2 seconds)
```

We could also report an error if `quit-if-one-screen=false` and `clear-screen-on-exit=false`, but this would be annoying if you tried to use `jj --config ui.streampager.quit-if-one-screen=false` and got an error.

-----------------

Here's the best option I have at this point:

```toml
[ui.streampager]
# Do not clear screen on exit. Use a full-screen interface for
# long output only. Like `less -FX`.
interface = "quit-if-one-page"  # (default).
# Always use a full-screen interface, ask the terminal to
# clear the screen on exit. Like `less +F +X`.
interface = "full-screen-clear-output"
# Use the alternate screen if the input is either long or takes
# more than `quit-quickly-delay-millis` to finish. Similar but not
# identical to `less -F +X`
interface = "quit-quickly-or-clear-output"

# Only relevant in the `quit-quickly-or-clear-output` mode
quit-quickly-delay-millis = 2000 # (defaults to 2 seconds)
```

I prefer this to the upstream's version (where the above options would correspond to, in order, "hybrid", "full-screen", and "delayed").

WDYT?

---

> Here's the best option I have at this point:

Looks good, but to be clear, I don't think it's any better than the streampager's naming. I'll probably need to read the documentation to understand the behavior anyway. (and I tend to use whatever the upstream provides in that case.)

fwiw, I would change the duration unit to "seconds" by default, and allow optional unit suffix or floating-point number.

---

> Looks good, but to be clear, I don't think it's any better than the streampager's naming.

Well, it's the best I could do. I think it's a little better than Streampager's.

> I would change the duration unit to "seconds" by default, and allow optional unit suffix or floating-point number.

This is a bit of a struggle. I could make it non-configurable for now.

Do we parse such suffixes anywhere?

One option is to do it manually and allow `ms` and `s` suffixes.

For floating points, I tried https://github.com/ilyagr/jj/commit/serde-float, but it results in a weird error when I try to use it with `--config`.

---

I made the duration non-configurable for now.

---

> I made the duration non-configurable for now.

Sounds good.

For parsing, we can use `config.get_value_with(name, parsing_fn)`. See `snapshot.max-new-file-size` handling.